### PR TITLE
[Examples] Improve the model card pushed from the `train_text_to_image.py` script

### PR DESCRIPTION
**Pull request by sayakpaul**
Currently, our training examples include limited information in the model cards. For example, one doesn't know how the code should be run with the model artifact(s) from the training examples just by taking a look at the corresponding model repository (such as [this one](https://huggingface.co/sayakpaul/lora-trained)).

We can improve this.  

So, to that end, this PR modifies the `train_text_to_image.py` to include a code snippet in the model card that we create from the example itself. Additionally, it includes the following things in the model card for a better developer experience which can be enabled directly from the Hub:

* Info on key training hyperparameters
* Dataset (and properly links that to the model card metadata)
* Weights and Biases run page when available (which discloses all the CLI args passed to the training script and details on the env such as accelerator type, etc.)

Here's how a pipeline repo looks like with the changes introduced in this PR: https://huggingface.co/sayakpaul/new_sd_pokemon.

The training was conducted with the following command:

```bash
export MODEL_NAME="CompVis/stable-diffusion-v1-4"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"

accelerate launch train_text_to_image.py   \
  --pretrained_model_name_or_path=$MODEL_NAME   \
  --dataset_name=$DATASET_NAME   \
  --mixed_precision="fp16"   \
  --resolution=512 --center_crop --random_flip   \
  --train_batch_size=1   \
  --gradient_accumulation_steps=4   \
  --gradient_checkpointing   \
  --max_train_steps=100   \
  --learning_rate=1e-05   \
  --max_grad_norm=1   --lr_scheduler="constant" --lr_warmup_steps=0   \
  --output_dir="sd-pokemon-model"   \
  --report_to="wandb"   \
  --validation_epochs=1 \
  --validation_prompts "cute dragon creature" "cute pokemon creature" "blue pokemon"   \
  --checkpointing_steps=50   \
  --output_dir="new_sd_pokemon"   \
  --push_to_hub
```

If there's a consensus, I can adapt these changes and include them in the rest of the scripts. Thanks to @osanseviero for the hint.

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me

File: examples/text_to_image/train_text_to_image.py [hunk 1]
```diff
@@ -62,6 +63,90 @@
...
+    images=None,
+    repo_folder=None,
+):
+    img_str = ""
+    if images is not None:
+        image_grid = make_image_grid(images, 1, len(args.validation_prompts))
+        image_grid.save(os.path.join(repo_folder, "val_imgs_grid.png"))
+        img_str += "![val_imgs_grid](./val_imgs_grid.png)\n"
+
+    yaml = f"""
```
Comment (osanseviero) [inline examples/text_to_image/train_text_to_image.py:89]: FYI `huggingface_hub` has nice utilities for model card creation (both content and metadata) and templates you could leverage here.

Comment (osanseviero) [reply examples/text_to_image/train_text_to_image.py:89]: See https://huggingface.co/docs/huggingface_hub/guides/model-cards

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -971,6 +1059,7 @@ def collate_fn(examples):
         pipeline.save_pretrained(args.output_dir)
 
         if args.push_to_hub:
+            save_model_card(args, repo_id, images, repo_folder=args.output_dir)
```
Comment (williamberman) [inline examples/text_to_image/train_text_to_image.py:1062]: Can you separately create the images that are saved to the model card here? Feel free to just pull the subset of code out of the `log_validation_images` that creates the images into a helper method.

File: examples/text_to_image/train_text_to_image.py [hunk 3]
```diff
@@ -62,6 +63,90 @@
...
+license: creativeml-openrail-m
+base_model: {args.pretrained_model_name_or_path}
+datasets:
+- {args.dataset_name}
+tags:
+- stable-diffusion
+- stable-diffusion-diffusers
+- text-to-image
+- diffusers
+inference: true
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:100]: I think this is the default

File: examples/text_to_image/train_text_to_image.py [hunk 4]
```diff
@@ -62,6 +63,90 @@
...
+---
+    """
+    model_card = f"""
+# Text-to-image finetuning - {repo_id}
+
+This pipeline was finetuned from **{args.pretrained_model_name_or_path}** on the **{args.dataset_name}** dataset. Below are some examples images that were generated with the finetuned pipeline (with the following prompts: {args.validation_prompts}): \n
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:106]:
```suggestion
This pipeline was finetuned from **{args.pretrained_model_name_or_path}** on the **{args.dataset_name}** dataset. Below are some example images generated with the finetuned pipeline using the following prompts: {args.validation_prompts}: \n
```

File: examples/text_to_image/train_text_to_image.py [hunk 5]
```diff
@@ -62,6 +63,90 @@
...
+
+pipeline = DiffusionPipeline.from_pretrained("{repo_id}", torch_dtype=torch.float16)
+prompt = "{args.validation_prompts[0]}"
+image = pipeline(prompt).images[0]
+image.save("my_image.png")
+```
+
+## Training info
+
+Following are the key hyperparameters that were used to run finetuning:
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:125]:
```suggestion
These are the key hyperparameters used during training:
```

File: examples/text_to_image/train_text_to_image.py [hunk 6]
```diff
@@ -62,6 +63,90 @@
...
+    if is_wandb_available():
+        wandb_run_url = None
+        if wandb.run is not None:
+            wandb_run_url = wandb.run.url
+
+    wandb_info = f"""
+More information on all the CLI arguments and the environment should be available on the `wandb` run page if you used it via `report_to="wandb"`.
+Check it out here: {wandb_run_url}.
+    """
+    model_card += wandb_info
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:144]: I would only present this if `wandb_run_url` is not `None`. Otherwise we'll have a "Check it out here: None"

File: examples/text_to_image/train_text_to_image.py (hunk 2 above)
Comment (pcuenca) [reply examples/text_to_image/train_text_to_image.py:1062]: Either that, or skip the image block in the model card. I'd prefer to have the images, but we may not be able to generate them unless we make up a prompt that may not be ideal for the particular fine-tune the user performed.

File: examples/text_to_image/train_text_to_image.py [hunk 7]
```diff
@@ -62,6 +63,90 @@
...
+## Training info
+
+Following are the key hyperparameters that were used to run finetuning:
+
+* Epochs: {args.num_train_epochs}
+* Learning rate: {args.learning_rate}
+* Batch size: {args.train_batch_size}
+* Gradient accumulation steps: {args.gradient_accumulation_steps}
+* Image resolution: {args.resolution}
+* Mixed-precision: {args.mixed_precision}
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:132]: I'd get this from the `accelerator` object instead, as this may be `None`. Or make sure we always populate `args.mixed_precision` with the value actually used after we instantiate `Accelerator`.

**Review by pcuenca — COMMENTED** [review]
This is a great improvement, just focused on some details. I agree with @osanseviero that we could maybe leverage the facilities in `huggingface_hub` (and we already have `Jinja2` as a requirement for training).

File: examples/text_to_image/train_text_to_image.py (hunk 2 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:1062]: If you see the logic in the new `save_model_card()` function, that is how it's currently done. 

If there are validation images to save to the model card, `img_str` will be crafted accordingly, otherwise, it will be none.

File: examples/text_to_image/train_text_to_image.py (hunk 3 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:100]: I thought making it explicit will be useful for the users.

File: examples/text_to_image/train_text_to_image.py (hunk 7 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:132]: Very good point! Let me do that.

File: examples/text_to_image/train_text_to_image.py (hunk 6 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:144]: Make sense. Let me do that.

File: examples/text_to_image/train_text_to_image.py [hunk 8]
```diff
@@ -747,8 +836,10 @@ def collate_fn(examples):
     weight_dtype = torch.float32
     if accelerator.mixed_precision == "fp16":
         weight_dtype = torch.float16
+        args.mixed_precision = accelerator.mixed_precision
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:839]: @pcuenca per your [suggestion](https://github.com/huggingface/diffusers/pull/3810/files#r1233108796).

File: examples/text_to_image/train_text_to_image.py [hunk 9]
```diff
@@ -62,6 +63,92 @@
...
+"""
+    if is_wandb_available():
+        wandb_run_url = None
+        if wandb.run is not None:
+            wandb_run_url = wandb.run.url
+
+    if wandb_run_url is not None:
+        wandb_info += f"Check it out here: {wandb_run_url}."
+
+    model_card += wandb_info
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image.py:146]: @pcuenca refactored the logic here.

File: examples/text_to_image/train_text_to_image.py (hunk 1 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:89]: Thanks for suggesting! 

But I feel like the current way of creating the content in the model card is a bit more explicit and also flexible. 

I could leverage something like https://huggingface.co/docs/huggingface_hub/guides/model-cards#from-a-jinja-template, but I would not here because:

* It doesn't reduce the code complexity significantly. 
* Now, users will have to learn about the model card related details of `huggingface_hub`, which I'd like to avoid because that's not the objective here. 

Cc: @pcuenca

**Comment by sayakpaul** [comment]
There's a 504 error coming from the HF Hub currently. Will try again later.

File: examples/text_to_image/train_text_to_image.py [hunk 10]
```diff
@@ -62,6 +63,92 @@
...
+* Learning rate: {args.learning_rate}
+* Batch size: {args.train_batch_size}
+* Gradient accumulation steps: {args.gradient_accumulation_steps}
+* Image resolution: {args.resolution}
+* Mixed-precision: {args.mixed_precision}
+
+"""
+    wandb_info = """
+More information on all the CLI arguments and the environment should be available on the `wandb` run page if you used it via `report_to="wandb"`.
+"""
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:137]:
```suggestion
    wandb_info = ""
```

Maybe put it later? (just a suggestion)

File: examples/text_to_image/train_text_to_image.py [hunk 11]
```diff
@@ -62,6 +63,92 @@
...
+    if is_wandb_available():
+        wandb_run_url = None
+        if wandb.run is not None:
+            wandb_run_url = wandb.run.url
+
+    if wandb_run_url is not None:
+        wandb_info += f"Check it out here: {wandb_run_url}."
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:144]:
```suggestion
        wandb_info = """
 More information on all the CLI arguments and the environment are available on your [`wandb` run page]({wandb_run_url}).
"""
```

File: examples/text_to_image/train_text_to_image.py [hunk 12]
```diff
@@ -935,12 +1026,13 @@ def collate_fn(examples):
                 break
 
         if accelerator.is_main_process:
+            images = None
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:1029]:
```suggestion
```

File: examples/text_to_image/train_text_to_image.py [hunk 13]
```diff
@@ -935,12 +1026,13 @@ def collate_fn(examples):
...
             if args.validation_prompts is not None and epoch % args.validation_epochs == 0:
                 if args.use_ema:
                     # Store the UNet parameters temporarily and load the EMA parameters to perform inference.
                     ema_unet.store(unet.parameters())
                     ema_unet.copy_to(unet.parameters())
-                log_validation(
+                images = log_validation(
```
Comment (pcuenca) [inline examples/text_to_image/train_text_to_image.py:1035]:
```suggestion
                log_validation(
```

**Review by pcuenca — APPROVED** [review]

**Comment by sayakpaul** [comment]
Model card with the latest details:

https://huggingface.co/sayakpaul/da-vinci-sd-pokemon

File: examples/text_to_image/train_text_to_image.py (hunk 2 above)
Comment (pcuenca) [reply examples/text_to_image/train_text_to_image.py:1086]: Yeah, that works for me.

File: examples/text_to_image/train_text_to_image.py (hunk 1 above)
Comment (Wauplin) [reply examples/text_to_image/train_text_to_image.py:89]: Thanks for the ping @osanseviero. I would also advocate to use `huggingface_hub`'s modelcards for a few reasons:
1. You need to be careful about encoding issues in the README (especially `\n` vs `\r\n` on Windows). It's quite specific and annoying but we did a few iterations in `hfh` to handle that correctly. The goal being to avoid big diffs if someone else updates the model card afterwards.
2. Using modelcards + a separate jinja template makes it really easy for non-developers to review the model card template without looking into the code. This can prove useful if someone from the ethics team (for example) wants to open a PR to complete the model card template without digging to the exact code (i.e. separate code and templates to separate usage). 
3. While it doesn't reduce much complexity, it doesn't add much either. I don't think users have to understand ModelCards internal details to use it correctly.

Comment (Wauplin) [reply examples/text_to_image/train_text_to_image.py:89]: That being said, I don't think the `Path('custom_template.md').write_text(template_text)` line [from the docs](https://huggingface.co/docs/huggingface_hub/guides/model-cards#from-a-jinja-template) should be reused. I think it would be best to provide a jinja template alongside the training script and only have:

```py
card_data = ModelCardData(language='en', license='mit', library_name='keras')
card = ModelCard.from_template(card_data, template_path='custom_template.md', author='nateraw')
card.save('README.md')
```

in the training script

Comment (Wauplin) [reply examples/text_to_image/train_text_to_image.py:89]: (thinking out loud here but) actually reusing the [default modelcard template](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/templates/modelcard_template.md) is also an option. It simplifies the example training script on your side (no need for a custom template) as the cost of a more verbose model card -with a lot of empty fields at first-. We can see this as a way to encourage users to document better their models (which is the goal of all of this process).

Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image.py:89]: I am fine having a more verbose model card but having to provide a Jinja template separately is something I am not comfortable doing for our examples. 

> Using modelcards + a separate jinja template makes it really easy for non-developers to review the model card template without looking into the code. This can prove useful if someone from the ethics team (for example) wants to open a PR to complete the model card template without digging to the exact code (i.e. separate code and templates to separate usage).

I think they would need to open up the PR editing the README file anyway whose content is straightforward IMO. 

Probably, the best is to reuse [default modelcard template](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/templates/modelcard_template.md) as you mentioned. Happy to accept a PR to see the changes.
